xdsl/transforms/memref_stream_legalize.py

```diff
@@ -0,0 +1,163 @@
+    f16,
+    f32,
+)
+from xdsl.dialects.linalg import IteratorType
+from xdsl.ir import Attribute, Operation
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.utils.exceptions import DiagnosticException
+from xdsl.utils.hints import isa
+
+StreamingVectorType: TypeAlias = VectorType[Float64Type | Float32Type | Float16Type]
+StreamingLegalType: TypeAlias = StreamingVectorType | Float64Type
+
+
+def _is_legal_vector(attr: Attribute) -> bool:
+    if isa(attr, StreamingVectorType):
+        match attr.element_type, attr.element_count():
+            case Float64Type(), 1:
+                return True
+            case Float32Type(), 2:
+                return True
+            case Float16Type(), 4:
+                return True
+            case _:
+                return False
+    return False
+
+
+def _is_legal_attr(attr: Attribute) -> bool:
+    return _is_legal_vector(attr) or isa(attr, StreamingLegalType)
+
+
+def _legalize_attr(attr: Attribute) -> StreamingLegalType:
+    if isa(attr, StreamingVectorType):
+        # Either already a legal vector or impossible to legalize
+        if not _is_legal_vector(attr):
+            raise DiagnosticException(f"Cannot legalize {attr} for streaming")
+        return attr
+    elif isinstance(attr, Float64Type):
+        return attr
+    elif isinstance(attr, Float32Type):
+        return VectorType(f32, (2,))
+    elif isinstance(attr, Float16Type):
+        return VectorType(f16, (4,))
+    else:
+        raise DiagnosticException(f"Cannot legalize {attr} for streaming")
+
+
+def _legalize_results(ops: set[Operation], rewriter: PatternRewriter) -> None:
+    if not ops:
+        return
+    new_ops: set[Operation] = set()
+    for use_op in ops:
+        illegal_results: list[int] = [
+            result.index for result in use_op.results if not _is_legal_attr(result.type)
+        ]
+        new_op = use_op.create(
+            operands=use_op.operands,
+            result_types=[_legalize_attr(res.type) for res in use_op.results],
+            attributes=use_op.attributes,
+        )
+        rewriter.replace_op(use_op, new_op)
+        for idx in illegal_results:
+            new_ops.update(use.operation for use in new_op.results[idx].uses)
+    _legalize_results(new_ops.difference(ops), rewriter)
+
+
+@dataclass(frozen=True)
+class MemrefStreamGenericLegalize(RewritePattern):
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(
+        self, op: memref_stream.GenericOp, rewriter: PatternRewriter
+    ) -> None:
+        if op.iterator_types.data[-1].data != IteratorType.PARALLEL:
+            raise DiagnosticException(
+                "iterators other than 'parallel' are not supported yet"
+            )
+        # Collect block arguments that need to be legalized
+        legalizations: dict[int, Attribute] = {
```

Comment:
It feels a little off to me to cast below. Maybe we're missing another helper that returns an optional `StreamingVectorType` or None of there's no legalization to do?
```suggestion
        legalizations: dict[int, StreamingVectorType] = {
```